MudCarousel: Fixed timer never being disposed;

Yomodo opened the pull request:
In my attempt to improve `MudCarousel` (#1913) I forgot to actually dispose the timer.

mikes-gh commented on src/MudBlazor/Components/Carousel/MudCarousel.razor.cs:
```diff
@@ -208,8 +208,13 @@ public async ValueTask DisposeAsync()
         protected virtual async ValueTask DisposeAsync(bool disposing)
         {
-            if (disposing)
+            if (disposing && null != _timer)
```
We prefer `_timer != null`. i.e. the null on the right side of the check 👍

henon commented on src/MudBlazor/Components/Carousel/MudCarousel.razor.cs:
```diff
@@ -208,8 +208,13 @@ public async ValueTask DisposeAsync()
         protected virtual async ValueTask DisposeAsync(bool disposing)
         {
-            if (disposing)
+            if (disposing && _timer != null)
+            {
                 await StopTimerAsync();
+
+                await _timer.DisposeAsync();
+                _timer = null;
+            }
```
The safest way to do this is to copy the reference in a local variable first, that way there can never be a race condition even if the method is called twice by different threads:

```c#
var timer=_timer;
if (timer!=null) {
  _timer=null;
  timer.DisposeAsync();
}
```

henon replied:
of course first do a `if (!disposing) return;` to potentially decouple from everything that will be disposed later.

henon replied:
or you simply do this:

```c#
if (!disposing) 
  return;
_timer?.DisposeAsync();
_timer=null;
```

henon reviewed (changes requested):
Thanks

Yomodo commented:
I had to move the timer instantiation so that the `DisposeAsync` method gets called, which was also not that case.

henon commented:
what if you move the disposing to Dispose instead of DisposeAsync? will it make a difference?

Yomodo commented:
Unfortunately does not matters, `sync`/`async`. I've tried all possible lifetime methods to see if that helps.
I've noticed that, even after page refresh, method `TimerTickAsync()` is being kept alive.

mikes-gh commented:
@Yomodo

When you refresh a page all the assets are re-downloaded and a new Blazor server connection is established.
Essentially you restart the app
You can see this by opening the console in dev-tools and observing the. unique ID of the Blazor server connection changing.,

Yomodo commented:
@mikes-gh Yes, my understanding also. Having read that I swapped browsers from Firefox to Microsoft Edge.
Lo and Behold, method `TimerTickAsync()` is **not** being kept alive! So it seems like a Firefox issue.

mikes-gh commented:
I don't think this is server side right ? so its likely dispose wont be called for 3 mins according to this

https://github.com/dotnet/aspnetcore/issues/18556

Yomodo commented:
I'm using the docs Carousel example server, so server side not wasm, right?
I'll give the monitoring a go.

mikes-gh commented:
Yes its both but yes likely you are using the server side version.  Maybe monitor the server for Blazor Disconnect message.  Maybe Edge/Chrome send this and Firefox doesnt on refresh

Yomodo commented:
Yes, Firefox is really the culprit here. 
My previous commit (https://github.com/Garderoben/MudBlazor/pull/2073/commits/5d816b8431abdca8e24bf0913c2206f5b9c27aef) was also not necessary and due to Firefox not doing something correctly.

Moving the time initialization code back to `OnAfterRenderAsync` works as expected with Microsoft Edge.

I'll change this back and will switch to Microsoft Edge as primary browser for development.

mikes-gh commented:
@Yomodo Please can you revert to minimum changes please. 👍

mikes-gh commented:
>TimerTickAsync() is not being kept alive!

How are you debugging this.  I would like to reproduce it.

Yomodo commented:
In `MudCarousel.razor.cs` I added: `private int count = 0;`
In method `TimerTickAsync()` I added: `System.Diagnostics.Debug.WriteLine($"{++count}: {nameof(TimerTickAsync)}");`
In method `DisposeAsync(bool disposing)` I added: `System.Diagnostics.Debug.WriteLine($"{count}: {nameof(DisposeAsync)}");`

- VS: `MudBlazor.Docs.Server` as startup and Firefox as the selected browser, start debug;
- Go to `Carousel` item in docs
- On page load this is visibile in the VS Output window:
1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
The timer has initialized for the 5 Carousels on the page. OK.

After 5 seconds:
2: TimerTickAsync
2: TimerTickAsync
2: TimerTickAsync
2: TimerTickAsync
2: TimerTickAsync
The timer elapsed for all 5 Carousels on the page. This goes on OK.
(Offtopic: Only 2 timers should tick because of `AutoCycle = true`)

-Click on the `Card` doc item, the timer dispose messages or all 5 Carousels should appear:
2: DisposeAsync
2: DisposeAsync
2: DisposeAsync
2: DisposeAsync
2: DisposeAsync

So far, all OK.

- Click on the `Carousel` item again and let it run a few ticks.
- Refresh the page to start it all again.

This time the VS Output window shows:
- A couple of `System.InvalidOperationException`'s
- No dispose messages;
- The counter 3 are the previous instances;

1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
1: TimerTickAsync
3: TimerTickAsync
3: TimerTickAsync
3: TimerTickAsync
3: TimerTickAsync
3: TimerTickAsync

henon commented:
Thank you very much!
